text/0000-restrict-constants-in-patterns.md

```diff
@@ -0,0 +1,608 @@
+    const B: bool = false;
+}    
+```
+
+Is this match exhaustive? Does it contain dead code? The answer will
+depend on whether `T=i32` or `T=u32`, of course.
+
+### Advantages of the current approach
+
+However, structural equality also has a number of advantages:
+
+**Better optimization.** One of the biggest "pros" is that it can
+potentially enable nice optimization. For example, given constants like the following:
+
+```rust
+struct Value { x: u32 }
+const V1: Value = Value { x: 0 };
+const V2: Value = Value { x: 1 };
+const V3: Value = Value { x: 2 };
+const V4: Value = Value { x: 3 };
+const V5: Value = Value { x: 4 };
+```
+
+and a match pattern like the following:
+
+```rust
+match v {
+    V1 => ..., 
+    ...,
+    V5 => ...,
+}
+```
+
+then, because pattern matching is always a process of structurally
+extracting values, we can compile this to code that reads the field
+`x` (which is a `u32`) and does an appropriate switch on that
+value. Semantic equality would potentially force a more conservative
+compilation strategy.
+
+**Better exhautiveness and dead-code checking.** Similarly, we can do
+more thorough exhaustiveness and dead-code checking. So for example if
+I have a struct like:
+
+```rust
+struct Value { field: bool }
+const TRUE: Value { field: true };
+const FALSE: Value { field: false };
+```
+
+and a match pattern like:
+
+```rust
+match v { TRUE => .., FALSE => .. }
+```
+
+then we can prove that this match is exhaustive. Similarly, we can prove
+that the following match contains dead-code:
+
+```rust
+const A: Value { field: true };
+match v {
+    TRUE => ...,
+    A => ...,
+}
+```
+
+Again, some of the alternatives might not allow this. (But note the
+cons, which also raise the question of exhaustiveness checking.)
+
+**Nullary variants and constants are (more) equivalent.** Currently,
+there is a sort of equivalence between enum variants and constants, at
+least with respect to pattern matching. Consider a C-like enum:
+
+```rust
+enum Modes {
+    Happy = 22,
+    Shiny = 44,
+    People = 66,
+    Holding = 88,
+    Hands = 110,
+}
+
+const C: Modes = Modes::Happy;
+```
+
+Now if I match against `Modes::Happy`, that is matching against an
+enum variant, and under *all* the proposals I will discuss below, it
+will check the actual variant of the value being matched (regardless
+of whether `Modes` implements `PartialEq`, which it does not here). On
+the other hand, if matching against `C` were to require a `PartialEq`
+impl, then it would be illegal. Therefore matching against an *enum
+variant* is distinct from matching against a *constant*.
+
+# Detailed design
+[design]: #detailed-design
+
+The goal of this RFC is not to decide between semantic and structural
+equality. Rather, the goal is to restrict pattern matching to that subset
+of types where the two variants behave roughly the same.
+
+### The structural match attribute
+
+We will introduce an attribute `#[structural_match]` which can be
```

Comment:
I aimed for strict structural equality, i.e. `==` is equivalent to `memcmp` (modulo padding).
It doesn't hold for references, two references containing different bit patterns can compare equal, just like e.g. floats.